src/QMCWaveFunctions/Fermion/DelayedUpdateCUDA.h

```diff
@@ -0,0 +1,205 @@
+                                          cudaMemcpyDeviceToHost, hstream), "cudaMemcpyAsync failed!");
+          prefetched_range.setRange(rowchanged, last_row);
+          waitStream();
+        }
+        // save AinvRow to new_AinvRow
+        std::copy_n(Ainv_buffer[prefetched_range.getOffset(rowchanged)], invRow.size(), invRow.data());
+        if ( delay_count > 0 )
+        {
+          const T cone(1);
+          const T czero(0);
+          const int norb = Ainv.rows();
+          const int lda_Binv = Binv.cols();
+          // multiply V (NxK) Binv(KxK) U(KxN) AinvRow right to the left
+          BLAS::gemv('T', norb, delay_count, cone, U.data(), norb, invRow.data(), 1, czero, p.data(), 1);
+          BLAS::gemv('N', delay_count, delay_count, cone, Binv.data(), lda_Binv, p.data(), 1, czero, Binv[delay_count], 1);
+          BLAS::gemv('N', norb, delay_count, -cone, V.data(), norb, Binv[delay_count], 1, cone, invRow.data(), 1);
+        }
+      }
+
+      // accept with the update delayed
+      template<typename VVT>
+      inline void acceptRow(Matrix<T>& Ainv, int rowchanged, const VVT& psiV)
+      {
+        // update Binv from delay_count to delay_count+1
+        const T cminusone(-1);
+        const T czero(0);
+        const int norb = Ainv.rows();
+        const int lda_Binv = Binv.cols();
+        std::copy_n(Ainv_buffer[prefetched_range.getOffset(rowchanged)], norb, V[delay_count]);
+        std::copy_n(psiV.data(), norb, U[delay_count]);
+        delay_list[delay_count] = rowchanged;
+        // the new Binv is [[X Y] [Z x]]
+        BLAS::gemv('T', norb, delay_count+1, cminusone, V.data(), norb, psiV.data(), 1, czero, p.data(), 1);
+        // x
+        T y = -p[delay_count];
+        for(int i=0; i<delay_count; i++)
+          y += Binv[delay_count][i] * p[i];
+        Binv[delay_count][delay_count] = y = T(1) / y;
+        // Y
+        BLAS::gemv('T', delay_count, delay_count, y, Binv.data(), lda_Binv, p.data(), 1, czero, Binv.data()+delay_count, lda_Binv);
+        // X
+        BLAS::ger(delay_count, delay_count, cminusone, Binv[delay_count], 1, Binv.data()+delay_count, lda_Binv, Binv.data(), lda_Binv);
+        // Z
+        for(int i=0; i<delay_count; i++)
+          Binv[delay_count][i] *= -y;
+        delay_count++;
+        // update Ainv when maximal delay is reached
+        if(delay_count==lda_Binv) updateInvMat(Ainv,false);
+      }
+
+      inline void updateInvMat(Matrix<T>& Ainv, bool transfer_to_host = true)
+      {
+        // update the inverse matrix
+        if( delay_count>0 )
+        {
+          const T cone(1);
+          const T czero(0);
+          const int norb=Ainv.rows();
+          const int lda_Binv=Binv.cols();
+          const T cminusone(-1);
+          cudaErrorCheck( cudaMemcpyAsync(U_gpu.data(), U.data(), norb*delay_count*sizeof(T), cudaMemcpyHostToDevice, hstream), "cudaMemcpyAsync failed!");
```

Comment:
Worth putting the formulae here or referring to file with the formulae

Reply:
I need to do this in later PR for an accurate pointer.